### PR TITLE
PoC - Logging, Authentication, generated clients for IAM, Secrets, Cosmos

### PR DESCRIPTION
**Pull request by mrnugget**
This is just a proof-of-concept for several requirements that we're after:

* custom logger
* authentication with a token
* custom user-agent
* auto-generating clients with swagger
* IAM client to login
* Secrets client
* Cosmos client

The most important file to look at is `example.org` which contains examples for

* how to login
* make authenticated requests
* create a secret
* fetch a secret
* install a package without options (marathon-lb)
* install a package with options (jenkins)

Noteworthy:
* the Swagger default template for clients has been changed so we don't have to pass in a token every time
* the `PackageOptions` struct has been manually created inside the `dcos/cosmos` folder (which means we are able to manually create helper methods inside the sub-clients)

The code is not clean, it's not tested, there is duplication, and it could and should be moved around a little bit, but it allows us to see what's possible :)

### CODE REVIEW

File: swagger/cosmos.yaml [hunk 1]
```diff
@@ -0,0 +1,938 @@
+swagger: '2.0'
```
Comment (bamarni) [inline swagger/cosmos.yaml]: Can we have the spec definitions in an `openapi` directory? As it's the new name for them.

File: dcos/client.go [hunk 2]
```diff
@@ -38,9 +53,100 @@ func NewClientWithOptions(httpClient *http.Client, config *Config) (*Client, err
...
+	cosmosRuntime.Consumers["application/vnd.dcos.service.update-response+json"] = cosmosRuntime.Consumers[runtime.JSONMime]
+	cosmosRuntime.Producers["application/vnd.dcos.service.update-request+json;charset=utf-8;version=v1"] = cosmosRuntime.Producers[runtime.JSONMime]
+	cosmosRuntime.Consumers["application/vnd.dcos.package.repository.add-response+json"] = cosmosRuntime.Consumers[runtime.JSONMime]
+	cosmosRuntime.Producers["application/vnd.dcos.package.repository.add-request+json;charset=utf-8;version=v1"] = cosmosRuntime.Producers[runtime.JSONMime]
+	cosmosRuntime.Consumers["application/vnd.dcos.package.repository.delete-response+json"] = cosmosRuntime.Consumers[runtime.JSONMime]
+	cosmosRuntime.Producers["application/vnd.dcos.package.repository.delete-request+json;charset=utf-8;version=v1"] = cosmosRuntime.Producers[runtime.JSONMime]
+	cosmosRuntime.Consumers["application/vnd.dcos.service.describe-response+json"] = cosmosRuntime.Consumers[runtime.JSONMime]
+	cosmosRuntime.Producers["application/vnd.dcos.service.describe-request+json;charset=utf-8;version=v1"] = cosmosRuntime.Producers[runtime.JSONMime]
```
Comment (bamarni) [inline dcos/client.go]: Can these be part of the spec and autogenerated?

File: dcos/client.go [hunk 3]
```diff
@@ -38,9 +53,100 @@ func NewClientWithOptions(httpClient *http.Client, config *Config) (*Client, err
...
 		HTTPClient: httpClient,
 		Config:     config,
-		UserAgent:  fmt.Sprintf("%s(%s)", ClientName, Version),
+		IAM:        iamClient,
+		Secrets:    secretsClient,
+		Cosmos:     cosmosClient,
 	}, nil
 }
+
+func (c *Client) Login(username, password string) (string, error) {
```
Comment (bamarni) [inline dcos/client.go]: Given that we have different login methods (username/password, service account, oidc/saml), in the future either we say uid/password is the default and keep it named `Login`, then we'll add eg. `ServiceLogin`, we could also make the arguments more generic instead.

File: generate_swagger.sh [hunk 4]
```diff
@@ -0,0 +1,27 @@
+#!/usr/bin/env bash
```
Comment (bamarni) [inline generate_swagger.sh]: Ideally this would be triggered by `go generate`

**Comment by mrnugget** [comment]
> Can you re-run CI? As it now runs golint we should make sure it passes before merging.

Rebased and pushed. CI now runs and fails due to missing comments. But I have to say, since it's just a proof of concept, I really don't want to merge this as it is 😉 It should rather serve as an example to see what's possible and how it would be possible and what could be done in the future

File: dcos/client.go (hunk 2 above)
Comment (mrnugget) [reply dcos/client.go]: That would be nice, but I only found this: https://github.com/go-openapi/runtime/issues/33

**Pull request merged**